src/nvim/eval/funcs.c

```diff
@@ -6314,6 +6314,99 @@ static void f_range(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   }
 }
 
+// Evaluate "expr" for readdir().
+static varnumber_T readdir_checkitem(typval_T *expr, const char *name)
+{
+  typval_T save_val;
+  typval_T rettv;
+  typval_T argv[2];
+  varnumber_T retval = 0;
+  bool error = false;
+
+  prepare_vimvar(VV_VAL, &save_val);
+  set_vim_var_string(VV_VAL, name, -1);
+  argv[0].v_type = VAR_STRING;
+  argv[0].vval.v_string = (char_u *)name;
+
+  if (eval_expr_typval(expr, argv, 1, &rettv) == FAIL) {
+    goto theend;
+  }
+
+  retval = tv_get_number_chk(&rettv, &error);
+  if (error) {
+    retval = -1;
+  }
+
+  tv_clear(&rettv);
+
+theend:
+  set_vim_var_string(VV_VAL, NULL, 0);
+  restore_vimvar(VV_VAL, &save_val);
+  return retval;
+}
+
+// "readdir()" function
+static void f_readdir(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+  typval_T *expr;
+  const char *path;
+  garray_T ga;
+  Directory dir;
+
+  tv_list_alloc_ret(rettv, kListLenUnknown);
+  path = tv_get_string(&argvars[0]);
+  expr = &argvars[1];
+  ga_init(&ga, (int)sizeof(char *), 20);
+
+  if (!os_scandir(&dir, path)) {
+    smsg(_(e_notopen), path);
+  } else {
+    for (;;) {
+      bool ignore;
+
+      path = os_scandir_next(&dir);
+      if (path == NULL) {
+        break;
+      }
+
+      ignore = (path[0] == '.'
+                && (path[1] == NUL || (path[1] == '.' && path[2] == NUL)));
+      if (!ignore && expr->v_type != VAR_UNKNOWN) {
+        varnumber_T r = readdir_checkitem(expr, path);
+
+        if (r < 0) {
+          break;
+        }
+        if (r == 0) {
+          ignore = true;
+        }
+      }
+
+      if (!ignore) {
+        ga_grow(&ga, 1);
+        ((char_u **)ga.ga_data)[ga.ga_len++] = vim_strsave((char_u *)path);
```

Comment:
`vim_strsave()` is defined as follows. Therefore, if you use `xsrdup()` instead of `vim_strsave()`, you don't need to cast the `path`.

```C
char_u *vim_strsave(const char_u *string)
  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC FUNC_ATTR_NONNULL_ALL
{
  return (char_u *)xstrdup((char *)string);
}
```

Ref. #459.

Reply:
Okay! Should I also replace `char_u` with `uint8_t`?

Reply:
> where only characters are used it's ok to use plain char

Therefore, I think `char` is fine.

Reply:
Thank you, I will do.

Reply:
I don't think you have to go that far.

Reply:
Okay, I will only replace `vim_strsave`.

Reply:
Okay :)

Reply:
Done! `uint8_t` It was only in one place, except for `argv[0].val.v_string` and `sort_strings()`.